File: test/corpus/propagation/fallible_ref_argument.exs
defmodule RustQ.Corpus.Propagation.FallibleRefArgument do
  @moduledoc "Propagate fallible values and borrow them for reference arguments."

  use RustQ.Meta

  alias RustQ.Type, as: R

  @spec decode(term()) :: R.nif_result(R.path(:Matrix))
  defrust decode(_term) do
    {:ok, Matrix.default()}
  end

  @spec consume(R.ref(R.path(:Matrix))) :: R.nif_result(R.unit())
  defrust consume(_matrix) do
    :ok
  end

  @spec run(term()) :: R.nif_result(R.unit())
  defrust run(term) do
    consume(decode(term))
  end
end

File: test/corpus/propagation/fallible_ref_argument.rs
fn decode<'a>(_term: Term<'a>) -> NifResult<Matrix> {
    Ok(Matrix::default())
}

fn consume(_matrix: &Matrix) -> NifResult<()> {
    Ok(())
}

fn run<'a>(term: Term<'a>) -> NifResult<()> {
    consume(&decode(term)?)
}
